pkg/resources/stage.go

```diff
@@ -0,0 +1,414 @@
+	reader.Comma = stageIDDelimiter
+	lines, err := reader.ReadAll()
+	if err != nil {
+		return nil, fmt.Errorf("Not CSV compatible")
+	}
+
+	if len(lines) != 1 {
+		return nil, fmt.Errorf("1 line per stage")
+	}
+	if len(lines[0]) != 3 {
+		return nil, fmt.Errorf("3 fields allowed")
+	}
+
+	stageResult := &stageID{
+		DatabaseName: lines[0][0],
+		SchemaName:   lines[0][1],
+		StageName:    lines[0][2],
+	}
+	return stageResult, nil
+}
+
+// Stage returns a pointer to the resource representing a stage
+func Stage() *schema.Resource {
+	return &schema.Resource{
+		Create: CreateStage,
+		Read:   ReadStage,
+		Update: UpdateStage,
+		Delete: DeleteStage,
+		Exists: StageExists,
+
+		Schema: stageSchema,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+	}
+}
+
+// CreateStage implements schema.CreateFunc
+func CreateStage(data *schema.ResourceData, meta interface{}) error {
+	db := meta.(*sql.DB)
+	name := data.Get("name").(string)
+	database := data.Get("database").(string)
+	schema := data.Get("schema").(string)
+
+	builder := snowflake.Stage(name, database, schema)
+
+	// Set optionals
+	if v, ok := data.GetOk("url"); ok {
+		builder.WithURL(v.(string))
+	}
+
+	if v, ok := data.GetOk("credentials"); ok {
+		builder.WithCredentials(v.(string))
+	}
+
+	if v, ok := data.GetOk("file_format"); ok {
+		builder.WithFileFormat(v.(string))
+	}
+
+	if v, ok := data.GetOk("copy_options"); ok {
+		builder.WithCopyOptions(v.(string))
+	}
+
+	if v, ok := data.GetOk("encryption"); ok {
+		builder.WithEncryption(v.(string))
+	}
+
+	if v, ok := data.GetOk("comment"); ok {
+		builder.WithComment(v.(string))
+	}
+
+	q := builder.Create()
+
+	err := DBExec(db, q)
+	if err != nil {
+		return errors.Wrapf(err, "error creating stage %v", name)
+	}
+
+	stageID := &stageID{
+		DatabaseName: database,
+		SchemaName:   schema,
+		StageName:    name,
+	}
+	dataIDInput, err := stageID.String()
+	if err != nil {
+		return err
+	}
+	data.SetId(dataIDInput)
+
+	return ReadStage(data, meta)
+}
+
+// ReadStage implements schema.ReadFunc
+func ReadStage(data *schema.ResourceData, meta interface{}) error {
+	db := meta.(*sql.DB)
+	stageID, err := stageIDFromString(data.Id())
+	if err != nil {
+		return err
+	}
+
+	dbName := stageID.DatabaseName
+	schema := stageID.SchemaName
+	stage := stageID.StageName
+
+	q := snowflake.Stage(stage, dbName, schema).Describe()
+	stageDesc, err := descStage(db, q)
+	if err != nil {
+		return err
+	}
+
+	err = data.Set("name", stage)
+	if err != nil {
+		return err
+	}
+
+	err = data.Set("database", dbName)
+	if err != nil {
+		return err
+	}
+
+	err = data.Set("schema", schema)
+	if err != nil {
+		return err
+	}
+
+	err = data.Set("url", stageDesc.url)
+	if err != nil {
+		return err
+	}
+
+	err = data.Set("aws_external_id", stageDesc.awsExternalID)
+	if err != nil {
+		return err
+	}
+
+	err = data.Set("snowflake_iam_user", stageDesc.snowflakeIamUser)
+	if err != nil {
+		return err
+	}
+
```

Comment:
IIUC some fields are omitted here because you can't read them back from snowflake via `DESCRIBE|SHOW STAGE`. Can you add a comment listing those and why they are omitted?

Reply:
Also can you check that there is no way to read those back? Sorry I don't know this part of snowflake well.

Reply:
Added `file_format`, `copy_options` and `comment`. The other two fields (`credentials` and `encryption`) are not available in the either the `SHOW STAGES` or `DESCRIBE STAGE` queries.